feat(lsp): add start() (fka start_or_attach)

mfussenegger opened the pull request:
Putting this up for discussion.

This would be an alternative/subset of https://github.com/neovim/neovim/pull/18506 that I think would be forward compatible with a future project system.

Until the project system is in place users would have to configure `FileType` autocommands themselves, for example to setup `vscode-json-languageserver` one would drop the following into `ftplugin/json.lua`:

```lua
vim.lsp.start_or_attach({
  cmd = {'vscode-json-language-server', '--stdio'},
  name = 'json-ls',
  root_dir = vim.fs.find({'.git'}, { upward = true }),
})
```

(Or setup a `au FileType json lua vim.lsp.start_or_attach { ... }`, or via the `nvim_create_autocmd` API)


This could then be extended in both directions:

- A future project system would provide a different `reuse_client` predicate (or we adapt the default implementation).
- A `lsp.config` would automatically setup the `FileType` autocmds on `lsp.config({})`; `start_or_attach` could also be extended/changed that if only `{ name = ... }` is present it would lookup the entries from `lsp.config` and merge them.

This will need https://github.com/neovim/neovim/pull/18583

mfussenegger committed:
feat(lsp): add a start_or_attach function

clason commented:
I think it would be useful if `start_or_attach` returned the `client` (or at least the `client_id`); this would make it easier to for follow-up configuration in the same `ftplugin` or autocommand.

kylo252 commented on runtime/lua/vim/lsp.lua:
```diff
@@ -662,6 +663,30 @@ function lsp.client()
+function lsp.start_or_attach(config, reuse_client)
+  reuse_client = reuse_client
+    or function(client)
+      return client.config.root_dir == config.root_dir and client.name == config.name
```
wouldn't it be better if this already uses the streamline approach of `workspace_folders`? since `root_dir` is more or less a utility wrapper around it, imo

https://github.com/neovim/nvim-lspconfig/blob/9e6bcf5a8915e8423d5cc7f82c5069c11272184d/lua/lspconfig/configs.lua#L195-L201

mfussenegger replied:
It would make the user setup more complicated for the "I have a single workspaceFolder" use-case and having this default implementation doesn't prevent a multi workspaceFolder use-case in the future.

As outlined in the PR description a future project extension could provide a different `reuse_client` function or we could even change the default implementation in a _mostly_ BWC way

It would also be possible to extend this later, so that if a user provides an alternative `reuse_client` function the `workspaceFolders` automatically get extended if the new config has a different `root_dir` (or different `workspaceFolders`.

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -954,6 +954,20 @@ start_client({config})                                *vim.lsp.start_client()*
                     not be fully initialized. Use `on_init` to do any actions
                     once the client has been initialized.
 
+start_or_attach({config}, {reuse_client})          *vim.lsp.start_or_attach()*
```
The name is awkward because it should just be a  parameter on `start_client` :)

mfussenegger replied:
I renamed it to just `start` because I wasn't happy with the `start_or_attach` either - a `maybe_start_and_attach` would've been more accurate but that's way too mouthful.
Now it has some overlap with `start_client` but assuming that we may make that private at some point in the future I think it's okayish.

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -954,6 +954,20 @@ start_client({config})                                *vim.lsp.start_client()*
+                Parameters: ~
+                    {config}        (table) Same configuration as documented
+                                    in |lsp.start_client()|
+                    {reuse_client}  nil|fun(client: table): boolean Function
```
so reuse_client=false is literally the same thing as start_client? we don't need a new function for this

mfussenegger replied:
It also attaches the current buffer. We'd probably have to add two parameters to `start_client`, `reuse_client` and `(auto)attach = true|false` to keep the existing behavior by default. And then users would have to add an explicit `attach = true` to their configuration to opt into the new behavior. I'm not sure that would be an improvement over this.

justinmk replied:
ok, your call.  I wonder though, if we had this feature from the beginning, would we make multiple functions for it or would it be a feature of start_client() ?

clason replied:
> I wonder though, if we had this feature from the beginning, would we make multiple functions for it or would it be a feature of start_client() ?

Probably not. It still might be better to add a new function that does everything right, and then deprecate the old functions. (The biggest consumer of the old functions is nvim-lspconfig, which will be made obsolete -- as far as code goes -- by this PR anyway.)

mfussenegger replied:
> I wonder though, if we had this feature from the beginning, would we make multiple functions for it or would it be a feature of start_client() ?

Probably a single function.
I gave this some more thought and came to the same conclusion as @clason

gpanders commented:
https://github.com/neovim/neovim/pull/18583 is now merged, hopefully that can unblock this.

mfussenegger committed:
fixup! feat(lsp): add a start_or_attach function

mfussenegger commented:
I renamed the method to `start` and wrapped the `reuse_clients` function into a `opts` parameter to have more flexibility for future additions.
The reason for that is that I already considered adding a `bufnr` to control which buffer to attach.

 Adding both these options into the existing `config` parameter felt wrong to me - as it would drift the `config` table of `start` / `start_client` further apart, which I think at this point would add more confusion.

Also added more docs and overhauled some of the lsp intro docs.

Not sure if this is already ready in this state. I think core should take care of a `client.notify('workspace/didChangeConfiguration', { settings = client.config.settings })` call in the initialization, so users don't have to be told to add that, but could also follow up on this afterwards.

gpanders commented:
New LSP intro docs look really good to me. 

>I think core should take care of a client.notify('workspace/didChangeConfiguration', { settings = client.config.settings }) call in the initialization, so users don't have to be told to add that,

+1

clason commented on runtime/doc/lsp.txt:
```diff
@@ -24,88 +24,84 @@ QUICKSTART                                              *lsp-quickstart*
+ensure they're only active if the there is a LSP client running. An example:
+>
+    vim.api.nvim_create_autocmd('LspAttach', {
+      group = lsp_group,
```
Minor nit: this is not copy-pasteable since the group is not created yet -- either create this explicitly beforehand, or leave out?

clason commented on runtime/doc/lsp.txt:
```diff
@@ -24,88 +24,84 @@ QUICKSTART                                              *lsp-quickstart*
+
+To use other LSP features like hover, rename, etc. you can setup some
+additional keymaps. It's recommended to setup them in a |LspAttach| autocmd to
+ensure they're only active if the there is a LSP client running. An example:
```
```suggestion
ensure they're only active if there is a LSP client running. An example:
```

clason commented on runtime/doc/lsp.txt:
```diff
@@ -800,6 +796,55 @@ set_log_level({level})                               *vim.lsp.set_log_level()*
+                command must be present in the `$PATH` environment variable or an absolute path to the executable.
+                Shell constructs like `~` are NOT expanded.
+
+                `root_dir` path to the project root. Language servers use this
+                information to discover metadata like the dependencies of your
+                project and they tend to index the contents within the project
+                folder. The example above uses |vim.fs.find| and
```
I'd suggest explicitly mentioning that `root_dir` serves two purposes: beside the mentioned one, it is _also_ used by Neovim to keep track of whether to start a new server or attach to a running one. 

(This is implicit in the initial paragraph, but I think it wouldn't hurt to be as explicit as possible here, given the confusion about `root_dir` we've seen...)

mfussenegger replied:
It's already mentioned in the first sentence, do you think it warrants repeating? I wasn't sure

clason replied:
Yes, that's exactly what I think. Doesn't have to be a novel; just making clear that it's not _just_ for Language servers.

clason replied:
On further thought, we should probably not further conflate the _server_ root dir and the _Neovim_ root dir (which is now handled explicitly by `reuse_client`). Maybe we should go even further in separating these concerns, but that shouldn't block this PR.

clason commented on runtime/doc/lsp.txt:
```diff
@@ -24,88 +24,84 @@ QUICKSTART                                              *lsp-quickstart*
+     vim.lsp.start({
+        name = 'my-server-name',
+        cmd = {'name-of-language-server-executable'},
+        root_dir = vim.fs.dirname(vim.fs.find('.git')[1]),
```
As `.git` has been the single most problematic root marker for lspconfig (especially for people having a `.git` in their home directory), maybe it would be better to replace this with a more clear-cut one (e.g., `compile-commands.json` or `.stylua.toml`) throughout this documentation?

kylo252 replied:
+1 for me on using something other than `.git`

but I'd go for

- `.clangd` instead of `compile_commands.json`, see https://clangd.llvm.org/config#files, _noevim's repo which has `compile_commands.json` under `build/`_
- `.luarc.json` or instead of `.stylua.toml`, see https://github.com/sumneko/lua-language-server/wiki/Setting

other worthy considerations

- `tsconfig.json` or `.eslintrc`
- `pyproject.toml`
- `go.mod`
- `cargo.toml`

kylo252 commented on runtime/lua/vim/lsp.lua:
```diff
@@ -662,6 +663,66 @@ function lsp.client()
+    end
+  config.name = config.name or (config.cmd[1] and vim.fs.basename(config.cmd[1])) or nil
+  local bufnr = api.nvim_get_current_buf()
+  for _, client in pairs(lsp.get_active_clients()) do
+    if reuse_client(client, config) then
+      lsp.buf_attach_client(bufnr, client.id)
+      return
+    end
+  end
+  local client_id = lsp.start_client(config)
+  lsp.buf_attach_client(bufnr, client_id)
```
this will make debugging a little easier, alternatively we can add more trace log entries in `lsp.buf_attach_client()`

```suggestion
  for _, client in pairs(lsp.get_active_clients()) do
    if reuse_client(client, config) then
       return lsp.buf_attach_client(bufnr, client.id)
    end
  end
  local client_id = lsp.start_client(config)
  return lsp.buf_attach_client(bufnr, client_id)
```

clason replied:
That doesn't really help, as `buf_attach_client` doesn't return anything (meaningful).

Instead (as mentioned above), `lsp.start` should return the `client_id` (if it did attach).

kylo252 replied:
why not make `buf_attach_client` return the `client_id` then instead of `true`?

mfussenegger committed:
fixup! fixup! feat(lsp): add a start_or_attach function

williamboman commented on runtime/doc/lsp.txt:
```diff
@@ -800,6 +794,58 @@ set_log_level({level})                               *vim.lsp.set_log_level()*
+                `name` is an arbitrary name for the LSP client. It should be
+                unique per language server.
+
+                `cmd` the command as list - used to start the language server. The
+                command must be present in the `$PATH` environment variable or an absolute path to the executable.
+                Shell constructs like `~` are NOT expanded.
```
I don't recall all the details, but I wonder if it'd be worthwhile documenting how Windows handles (or rather doesn't handle) certain executables [through libuv](https://github.com/libuv/libuv/blob/2a1b880f5439e074ef4d5c556f626b1044fb6781/src/win/process.c#L304-L308). A large portion of servers are distributed as `.cmd`, `.ps1` or `.bat` scripts on Windows, which would not play nicely here as libuv wouldn't recognize them as executable. This is currently handled by lspconfig by using `cmd.exe /C` as an intermediary for launching server processes.

I think this could potentially be a source of high volume issues for Windows users.

mfussenegger replied:
I think that's out of scope of this PR

williamboman replied:
I was just thinking if this is a new API that is intended to be used directly by users it might lead to some hard-to-troubleshoot issues. I'd wager a vast majority of users would expect ergonomics similar to their shell environment, with path expansion and `PATHEXT` to be respected.

Just an example from the perspective of Windows users, the following would not function:

```lua
vim.lsp.start {
  cmd = { "typescript-language-server", "--stdio" },
  ..
}
```

Assuming you have your npm PATH set up correctly; jumping into a shell and typing `where typescript-language-server` would look fine, going into Neovim and trying `:echo executable("typescript-language-server")` would [also look fine](https://github.com/neovim/neovim/blob/c632f64e247c672e425f609bb47a9ab0517a4c31/src/nvim/os/fs.c#L323). I think this has been discussed before, but I think spawning RPC processes via a shell should be considered as more LSP functionality finds its way into core (or some other alternative to meet user expectations). There might be some other future plans that I'm not aware of so apologies if I'm overreaching

clason replied:
It is a new API that is intended to be used by moderately advanced users. New users (especially on Windows) will likely still rely on a lspconfig or lspinstaller plugin (which hopefully can be simplified using this new API).

Supporting "shell ergonomics for everyone" is _far_ outside the scope of this PR. (Although it may be reasonable to improve `vim.fs` to make things somewhat easier.)

williamboman commented on runtime/lua/vim/lsp.lua:
```diff
@@ -662,6 +663,69 @@ function lsp.client()
+---                            The default implementation re-uses a client if name
+---                            and root_dir matches.
+---@return number client_id
+function lsp.start(config, opts)
```
I have very little familiarity with Neovim's Lua module structure, but I feel like methods that only ever operate on the current buffer probably would be better put under `vim.lsp.buf`, no?

mfussenegger replied:
No

mfussenegger committed:
fixup! fixup! fixup! feat(lsp): add a start_or_attach function

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -662,6 +663,77 @@ function lsp.client()
+--- })
+--- </pre>
+---
+--- See |lsp.start_client| for all available options. The most important are:
```
should we deprecate start_client and move its docs to start()?

williamboman commented on runtime/doc/lsp.txt:
```diff
@@ -24,88 +24,82 @@ QUICKSTART                                              *lsp-quickstart*
+     vim.lsp.start({
+        name = 'my-server-name',
+        cmd = {'name-of-language-server-executable'},
+        root_dir = vim.fs.dirname(vim.fs.find({'setup.py', 'pyproject.toml'}, { upward = true })[1]),
```
Having it be a function would also allow potentially running it in async mode, making things like shelling out to locate an appropriate root dir less disruptive to the UX

mfussenegger replied:
> should this be a function so that it's eval'd dynamically?

I'm not sure if it is needed, given that `vim.lsp.start` will likely be called within a FileType event most of the time, so you already get some lazy evaluation due to that.

If we end up adding `lsp.config` too, that may change things.

> Having it be a function would also allow potentially running it in async mode, making things like shelling out to locate an appropriate root dir less disruptive to the UX

Only if the function supported a callback.
